documentation/design.md

```diff
@@ -0,0 +1,23 @@
+Config files API is a library written in ruby intended for accessing various configuration files. It is structured into several layers and creates an internal abstraction of configuration file.
+
+## First layer ##
+
+Is responsible for accessing configuration files itself. In simplest case it accesses local coinfiguration files, but it can be adapted to access remote, chrooted or memory files too.
+
+## Second layer ##
+
+This layer parses the configuration file which was loaded by the underlying layer. It knows structure of the file and transforms it into abstract representation. The library typically uses external tool for parsing here like Augeas. So, if the external tools has a specific requirements, it has to be satisfied to get things working. For example if the Augeas is in use you neet to provide him with a proper lens to parse the particular configuration file.
+
+## Third layer ##
+
+Last layer creates a model of the configuration file - an API for accessing configuration from an application. It basically creates "an abstraction on top of another abstraction". It means that is unifies use of various tools which can be used for accessing and parsing various configuration files.
+
+## Limitations ##
+
+The approach as described above has several limitations.
+
+Firstly you need to feed a parser too (e.g. provide lenses for augeas as described above). So, you at least has to know which parser is used for parsing which file and provide him with all stuff needed for parsing of file of your interest. This is especially important in case of nonstandard / custom configuration files. However this is limitation mainly for developer. If the developer plans to use CFA, then he has to evaluate if some work in this area has to be done as well.
+
+Second example of limitation is that the library and / or parsers on second layer use an abstraction for representing configuration file. This abstraction transform a configuration file into a model and establishes a relation between the file and the model. This relation is not bijective (typicall example is Augeas). It means that some irrelevant pieces of the configuration file are not represented in the model. For example some spaces can be left out if these are not needed from syntactic point of view. This can lead to loose of custom padding. Another example can be comments. You can often see that, if your file uses e.g. "#" as a comment mark, then some parsers can squash lines full of these marks (which some developers use as a kind of delimiter) to just one "#".
```

Comment:
in fact I do not see in augeas this squashing, but difference is that some lenses "eats" first '#' and some not. So some comments start with it and some not. so if you have comment "## my comment"

1. it can be represented as "## my comment"
2. or as "# my comment" if comment character is eaten

Reason for it is, that in some config files two different comment markers are allowed like "#" or ";", so real value is stored and not thrown away. But of course it can also eat all comment characters, it depends on lens.